src/mz-debug/src/internal_http_dumper.rs

```diff
@@ -0,0 +1,313 @@
+use reqwest::header::{HeaderMap, HeaderValue};
+use std::path::Path;
+use tokio::fs::{File, create_dir_all};
+use tokio::io::AsyncWriteExt;
+use tracing::{info, warn};
+
+use crate::kubectl_port_forwarder::{
+    KubectlPortForwarder, find_cluster_services, find_environmentd_service,
+};
+use crate::{Context, EmulatorContext, SelfManagedContext};
+
+static PROM_METRICS_ENDPOINT: &str = "metrics";
+static ENVD_HEAP_PROFILE_ENDPOINT: &str = "prof/heap";
+static CLUSTERD_HEAP_PROFILE_ENDPOINT: &str = "heap";
+static INTERNAL_HOST_ADDRESS: &str = "127.0.0.1";
+static INTERNAL_HTTP_PORT: i32 = 6878;
+
+/// A struct that handles downloading and saving profile data from HTTP endpoints
+pub struct InternalHttpDumpClient<'n> {
+    context: &'n Context,
+    http_client: &'n reqwest::Client,
+}
+
+/// A struct that handles downloading and exporting data from our internal HTTP endpoints.
+impl<'n> InternalHttpDumpClient<'n> {
+    pub fn new(context: &'n Context, http_client: &'n reqwest::Client) -> Self {
+        Self {
+            context,
+            http_client,
+        }
+    }
+
+    async fn dump_request_to_file(
+        &self,
+        relative_url: &str,
+        headers: HeaderMap,
+        output_path: &Path,
+    ) -> Result<(), anyhow::Error> {
+        // Try HTTPS first, then fall back to HTTP if that fails
+        let mut url = format!("https://{}", relative_url);
+        let mut response = self
+            .http_client
+            .get(&url)
+            .headers(headers.clone())
+            .send()
+            .await;
+
+        if response.is_err() {
+            // Fall back to HTTP if HTTPS fails
+            url = format!("http://{}", relative_url);
+            response = self.http_client.get(&url).headers(headers).send().await;
+        }
+
+        let response = response.with_context(|| format!("Failed to send request to {}", url))?;
+
+        if !response.status().is_success() {
+            return Err(anyhow::anyhow!(
+                "Failed to get response from {}: {}",
+                url,
+                response.status()
+            ));
+        }
+
+        let mut file = File::create(output_path)
+            .await
+            .with_context(|| format!("Failed to create file: {}", output_path.display()))?;
+
+        let mut stream = response.bytes_stream();
+        while let Some(chunk) = stream.next().await {
+            let chunk = chunk.with_context(|| "Failed to read chunk from response")?;
+            file.write_all(&chunk)
+                .await
+                .with_context(|| "Failed to write chunk to file")?;
+        }
+
+        file.flush()
+            .await
+            .with_context(|| "Failed to flush file contents")?;
+
+        Ok(())
+    }
+
+    /// Downloads and saves heap profile data
+    pub async fn dump_heap_profile(&self, relative_url: &str, service_name: &str) -> Result<()> {
+        let output_dir = self.context.base_path.join("profiles");
```

Comment:
Maybe these directory names should be constants that are grouped at the top of the file? That way it's easier to determine what folders the debug tool might create? Not blocking but just a thought

Reply:
Ah like saving `profiles` or `self.context.base_path.join("profiles")`? I think saving paths like `profiles` in static str variables like `RELATIVE_DIR` makes total sense.

The latter might be an issue since `base_path` has format `mz-debug_<DateTime>` where `DateTime` gets computed on `main` at the start of the program. We could have it done as a constant then have `internal_http_dumper` and everything alike refer to it, but I like the flexibility of having the DateTime be computed on runtime in case we:
a) Do multiple runs with multiple date times (i.e. `mz-debug/<DateTime>`)
b) We do some initialization that takes a while and wanna compute `DateTime` later